lxd/main_forkproxy.go

```diff
@@ -0,0 +1,266 @@
+		if (dosetns(connect_pid, "net") < 0) {
+			fprintf(stderr, "Failed setns to connector network namespace: %s\n", strerror(errno));
+			_exit(1);
+		}
+	}
+}
+*/
+import "C"
+
+type cmdForkproxy struct {
+	cmd    *cobra.Command
+	global *cmdGlobal
+}
+
+func (c *cmdForkproxy) Command() *cobra.Command {
+	// Main subcommand
+	cmd := &cobra.Command{}
+	cmd.Use = "forkproxy <listen PID> <listen address> <connect PID> <connect address> <fd> <reexec> <log path> <pid path>"
+	cmd.Short = "Setup network connection proxying"
+	cmd.Long = `Description:
+  Setup network connection proxying
+
+  This internal command will spawn a new proxy process for a particular
+  container, connecting one side to the host and the other to the
+  container.
+`
+	cmd.RunE = c.Run
+	cmd.Hidden = true
+
+	c.cmd = cmd
+	return cmd
+}
+
+func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
+	if len(args) != 8 {
+		cmd.Help()
+
+		if len(args) == 0 {
+			return nil
+		}
+
+		return fmt.Errorf("Missing required arguments")
+	}
+
+	// Get all our arguments
+	listenPid := args[0]
+	listenAddr := args[1]
+	connectPid := args[2]
+	connectAddr := args[3]
+
+	fd := -1
+	if args[4] != "0" {
+		fd, _ = strconv.Atoi(args[4])
+	}
+
+	// At this point we have already forked and should set this flag to 1
+	args[5] = "1"
+
+	// Check where we are in initialization
+	if !shared.PathExists(fmt.Sprintf("/proc/self/fd/%d", fd)) {
+		fmt.Printf("Listening on %s in %s, forwarding to %s from %s\n", listenAddr, listenPid, connectAddr, connectPid)
+
+		file, err := getListenerFile(listenAddr)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		listenerFd := file.Fd()
+		if err != nil {
+			return fmt.Errorf("failed to duplicate the listener fd: %v", err)
+		}
+
+		newFd, err := syscall.Dup(int(listenerFd))
+		if err != nil {
+			return fmt.Errorf("failed to dup fd: %v", err)
+		}
+
+		fmt.Printf("Re-executing proxy process\n")
+
+		args[4] = strconv.Itoa(int(newFd))
+		execArgs := append([]string{"lxd", "forkproxy"}, args...)
+
+		err = syscall.Exec(util.GetExecPath(), execArgs, os.Environ())
+		if err != nil {
+			return fmt.Errorf("failed to re-exec: %v", err)
+		}
+	}
+
+	// Re-create listener from fd
+	listenFile := os.NewFile(uintptr(fd), "listener")
+	listener, err := net.FileListener(listenFile)
+	if err != nil {
+		return fmt.Errorf("failed to re-assemble listener: %v", err)
+	}
+
+	defer listener.Close()
+
+	fmt.Printf("Starting to proxy\n")
+
+	// begin proxying
+	for {
+		// Accept a new client
+		srcConn, err := listener.Accept()
+		if err != nil {
+			fmt.Printf("error: Failed to accept new connection: %v\n", err)
+			continue
+		}
+		fmt.Printf("Accepted a new connection\n")
+
+		// Connect to the target
+		dstConn, err := getDestConn(connectAddr)
+		if err != nil {
+			fmt.Printf("error: Failed to connect to target: %v\n", err)
+			srcConn.Close()
+			continue
+		}
+
+		go io.Copy(srcConn, dstConn)
+		go io.Copy(dstConn, srcConn)
```

Comment:
This looks like it should be wrapped by our `EAGAIN` reader/writer interfaces I added given that there'll be a lot of signals floating around in the Go runtime.

Reply:
Indeed, added a commit for that one.